irods_specific_implementation/irods_rest_access_implementation.h

```diff
@@ -1,147 +1,168 @@
 
+                using json = nlohmann::json;
+
+                const auto results = json::object({
+                    {"headers", json::object({
+                        {"X-API-KEY", ticket_id}
+                    })},
+                    {"url", fmt::format("{}/stream?path={}&offset=0&limit={}", _base_url, logical_path, size)}
+                });
+
+                logger_->debug("Result = [{}]", results.dump());
+
+                return std::forward_as_tuple(Pistache::Http::Code::Ok, results.dump());
+            }
+            catch (const fs::filesystem_error& _e) {
+                auto error = make_error(_e.code().value(), _e.what());
+                return std::forward_as_tuple(Pistache::Http::Code::Bad_Request, error);
+            }
+            catch (const irods::exception& _e) {
+                auto error = make_error(_e.code(), _e.what());
+                return std::forward_as_tuple(Pistache::Http::Code::Bad_Request, error);
+            }
+            catch (const std::exception& _e) {
+                auto error = make_error(SYS_INTERNAL_ERR, _e.what());
+                return std::forward_as_tuple(Pistache::Http::Code::Bad_Request, error);
+            }
+        } // operator()
+
+    private:
+        std::string make_ticket_id() const
+        {
+            constexpr int ticket_len = 15;
+
+            // random_bytes must be (unsigned char[]) to guarantee that following
+            // modulo result is positive (i.e. in [0, 61])
+            unsigned char random_bytes[ticket_len];
+            irods::getRandomBytes(random_bytes, ticket_len);
+
+            const char character_set[] = {'A', 'B', 'C', 'D', 'E', 'F', 'G',
+            'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
+            'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g',
+            'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't',
+            'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6',
+            '7', '8', '9'};
+
+            std::string new_ticket;
+            for (int i = 0; i < ticket_len; ++i) {
+                const int ix = random_bytes[i] % sizeof(character_set);
+                new_ticket += character_set[ix];
+            }
+
+            return new_ticket;
+        } // make_ticket_id
+
+        void create_ticket(connection_proxy& _conn,
+                           const std::string_view _ticket_id,
+                           const std::string_view _logical_path) const
+        {
+            rx_ticket(*_conn(), "create", _ticket_id, "read", _logical_path);
+        } // create_ticket
+
+        void set_ticket_use_count(connection_proxy& _conn,
+                                  const std::string_view& _ticket_id,
+                                  const Pistache::Optional<std::string>& _use_count) const
+        {
+            const auto use_count = _use_count.getOrElse("1");
```

Comment:
this should default to zero ("0") (i.e. unlimited).

this matches AWS signed URL (canned policy of unlimited use and an expires time).

Reply:
done